src/PartitionLoops.cpp

```diff
@@ -217,16 +217,50 @@ struct Simplification {
     Interval interval;
 };
 
+template<typename T>
+class ExprUsesValidBuffers : public IRGraphVisitor {
+    using IRGraphVisitor::visit;
+
+    const Scope<T> &allocated_buffers;
+
+    void visit(const Load *op) {
+        if (!allocated_buffers.contains(op->name)) {
+            valid = false;
+        } else {
+            IRGraphVisitor::visit(op);
+        }
+    }
+public:
+    ExprUsesValidBuffers(const Scope<T> &buffers) : allocated_buffers(buffers), valid(true) {}
+    bool valid;
+};
+
+/** Check if any references to buffers in an expression is valid (i.e. the buffers
+ * have already been allocated).
+ */
+template<typename T>
+bool expr_uses_valid_buffers(Expr e, const Scope<T> &buffers) {
+    ExprUsesValidBuffers<T> uses(buffers);
+    e.accept(&uses);
+    return uses.valid;
+}
+
 // Then we define the visitor that hunts for them.
 class FindSimplifications : public IRVisitor {
     using IRVisitor::visit;
 
     Scope<int> depends_on_loop_var;
+    const Scope<int> &allocated_buffers;
 
     void new_simplification(Expr condition, Expr old, Expr likely_val, Expr unlikely_val) {
         if (!expr_uses_vars(condition, depends_on_loop_var)) {
             return;
         }
+        if (!expr_uses_valid_buffers(condition, allocated_buffers)) {
```

Comment:
Won't this discard conditions that load from input images? I think it might be better to build a scope of allocations to *not* use by adding an Allocate/Free visitor to FindSimplifications instead of PartitionLoops.